### PR TITLE
feat: LT-1190 End User Documents

### PR DESCRIPTION
**Pull request by markm5A8Z9F43913A**

### CODE REVIEW

File: test_helpers/clients.py [hunk 1]
```diff
@@ -189,6 +190,34 @@ def create_good_document(self, good: Good, user: ExporterUser, name: str, s3_key
         good_doc.save()
         return good_doc
 
+    @staticmethod
+    def create_document_for_end_user(end_user: EndUser):
```
[the remark this answers is not in the document]
Comment (markm5A8Z9F43913A) [reply test_helpers/clients.py]: Done.

File: applications/tests/tests_create.py [hunk 2]
```diff
@@ -47,3 +47,42 @@ def test_that_cannot_submit_with_no_sites_or_external(self):
 
         response = self.client.post(self.url, data, **self.exporter_headers)
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    # if POST - end user with no document - 400
+    def test_status_code_post_no_end_user_document(self):
+        # assemble
+        draft = self.create_standard_draft_without_end_user_document(self.exporter_user.organisation, 'test')
```
Comment (jfaracik) [inline applications/tests/tests_create.py]: Use self.organisation as opposed to self.exporter_user.organisation

File: cases/tests/tests_case_end_user_documents.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+from cases.models import Case
+from test_helpers.clients import DataTestClient
+
+
+class CaseEndUserDocumentTests(DataTestClient):
+    def setUp(self):
+        super().setUp()
+
+    def test_case_contains_end_user_document(self):
```
Comment (efahy) [inline cases/tests/tests_case_end_user_documents.py]: Comment explaining what's being tested please

File: end_user/document/tests.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.draft_no_user = self.create_draft(self.exporter_user.organisation, 'Dafty daft')
+        self.url_no_user = reverse('drafts:end_user_document', kwargs={'pk': self.draft_no_user.id})
+
+        self.data = {"name": "file123.pdf",
+                 "s3_key": "file123_12345678.pdf",
+                 "size": 476,
+                 "description": "Description 58398"}
+
+    # if POST - end-user set - GET returns correct data
```
Comment (efahy) [inline end_user/document/tests.py]: Docstring comments please

File: end_user/document/tests.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        # act
+        response = self.client.get(self.url_draft_with_user, **self.exporter_headers)
+
+        # assert
+        self.assertEqual(status.HTTP_200_OK, response.status_code)
+        self.assertEqual(None, response.json()['document'])
+
+    # if POST - document not set - return 201
+    @mock.patch('documents.tasks.prepare_document.now')
+    def test_status_code_post_document_not_exist(self, mock_obj):
```
Comment (efahy) [inline end_user/document/tests.py]: Please add another test below this to test that after a successful POST, the document has been created

File: end_user/document/tests.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    def test_status_code_delete_document_not_exist(self):
+        # act
+        response = self.client.delete(self.url_draft_with_user, **self.exporter_headers)
+
+        # assert
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    # if POST - document exist - return 400
+    @mock.patch('documents.tasks.prepare_document.now')
+    def test_status_code_post_document_exists(self, mock_obj):
```
Comment (efahy) [inline end_user/document/tests.py]: Extend this test to ensure there is still only one doc after the post or put the check in a similar, new test

File: end_user/document/tests.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        # act
+        response = self.client.post(self.url_draft_with_user, data=self.data, **self.exporter_headers)
+
+        # assert
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    # if DELETE - document exist - return 204
+    @mock.patch('documents.tasks.prepare_document.now')
+    @mock.patch('documents.models.Document.delete_s3')
+    def test_status_code_delete_document_exists(self, delete_s3_mock, prepare_document_now_mock):
```
Comment (efahy) [inline end_user/document/tests.py]: Test here or in another test that the doc is gone

File: end_user/serializers.py [hunk 8]
```diff
@@ -22,4 +26,31 @@ class Meta:
...
+
+    class Meta:
+        model = EndUserDocument
+        fields = ('id', 'name', 's3_key', 'size', 'end_user', 'description', 'safe')
+
+    def create(self, validated_data):
+        end_user_document = super(EndUserDocumentSerializer, self).create(validated_data)
+        end_user_document.save()
+
+        if BACKGROUND_TASK_ENABLED:
```
Comment (efahy) [inline end_user/serializers.py]: This if/else is now repeated in three places or at least there is code in 3 places that looks very similar!  Can we get a document helper method that we reuse everywhere or are there reasons why not?

File: end_user/document/models.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from django.db import models
+
+from documents.models import Document
+from end_user.models import EndUser
+
+
+class EndUserDocument(Document):
+    end_user = models.ForeignKey(EndUser, on_delete=models.CASCADE)
+    description = models.TextField(default=None, blank=True, null=True, max_length=280)
```
Comment (efahy) [inline end_user/document/models.py]: description seems like a field that would go better on the base document model... why are we restricting it to only End User Documents?  I think I'd move it and allow it to be null even if its not currently set for other document types

File: drafts/views/draft_end_user_document_views.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+from end_user.document.models import EndUserDocument
+from end_user.serializers import EndUserDocumentSerializer
+
+
+class EndUserDocuments(APIView):
+    authentication_classes = (ExporterAuthentication,)
+
+    def get(self, request, pk):
+        """
+        Returns a list of documents on the specified end user
```
Comment (efahy) [inline drafts/views/draft_end_user_document_views.py]: There can be only one, right?

File: drafts/views/draft_end_user_document_views.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        """
+
+        draft = get_draft(pk)
+        end_user = draft.end_user
+
+        if not end_user:
+            return JsonResponse(data={'error': 'No such user'},
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        end_user_document = EndUserDocument.objects.filter(end_user=end_user).first()
```
Comment (efahy) [inline drafts/views/draft_end_user_document_views.py]: Don't need to do the `.first()` in this case.  If there are any you'll be throwing the error.  Put an `s` on the variable name

File: drafts/views/draft_end_user_document_views.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        end_user = draft.end_user
+        if end_user is None:
+            return JsonResponse(data={'error': 'No such user'},
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        end_user_document = EndUserDocument.objects.filter(end_user=end_user).values()
+        return JsonResponse({'document': end_user_document[0] if end_user_document else None})
+
+
+    @swagger_auto_schema(
```
Comment (efahy) [inline drafts/views/draft_end_user_document_views.py]: PEP8 says too many blank lines.. same below

Comment (efahy) [reply drafts/views/draft_end_user_document_views.py]: (red underlines FTW)

File: applications/creators.py [hunk 13]
```diff
@@ -16,6 +17,12 @@ def create_standard_licence(draft, application, errors):
     if not draft.end_user:
         errors['end_user'] = get_string('applications.standard.no_end_user_set')
 
+    end_user_document = EndUserDocument.objects.filter(end_user=draft.end_user).first()
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/creators.py]: Couldn't agree more. I believe we may still be do something similar in another place as will so I will fix now

File: end_user/serializers.py (hunk 8 above)
Comment (BenAndrew-Kainos) [reply end_user/serializers.py]: Great spot. Have refactored this into a method called process_document now

File: end_user/document/tests.py (hunk 5 above)
Comment (markm5A8Z9F43913A) [reply end_user/document/tests.py]: test_correct_data_get_document

File: end_user/document/tests.py (hunk 7 above)
Comment (markm5A8Z9F43913A) [reply end_user/document/tests.py]: test_status_code_get_document_not_exist

File: applications/tests/tests_create.py [hunk 14]
```diff
@@ -41,9 +41,79 @@ def test_create_application_with_invalid_id(self):
         self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
 
     def test_that_cannot_submit_with_no_sites_or_external(self):
+        """
+        TODO: Add description
```
Comment (Mark-Spiers-kainos) [inline applications/tests/tests_create.py]: Please do

File: documents/__init__.py [hunk 15]
```diff
@@ -0,0 +1,2 @@
+def tasks():
+    return None
```
Comment (Mark-Spiers-kainos) [inline documents/__init__.py]: no newline

**Pull request merged**